src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,281 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
   zero_mem' := ⟨0, ⟨0, (mk_zero _).symm⟩⟩,
   neg_mem' := λ x ⟨n, ⟨a, h⟩⟩, h.symm ▸ ⟨n, ⟨-a, neg_mk _ _⟩⟩ }
 
+end
+
 local notation `A⁰_` f_deg := degree_zero_part f_deg
 
-instance (f : A) {m : ℕ} (f_deg : f ∈ 𝒜 m) : comm_ring (degree_zero_part f_deg) :=
+section
+
+variable {𝒜}
+
+instance (f : A) {m : ℕ} (f_deg : f ∈ 𝒜 m) : comm_ring (A⁰_ f_deg) :=
 (degree_zero_part f_deg).to_comm_ring
 
 /--
 Every element in the degree zero part of `Aₓ` can be written as `a/x^n` for some `a` and `n : ℕ`,
 `degree_zero_part.deg` picks this natural number `n`
 -/
-def degree_zero_part.deg {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) : ℕ :=
+def degree_zero_part.deg {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) : ℕ :=
 x.2.some
 
 /--
 Every element in the degree zero part of `Aₓ` can be written as `a/x^n` for some `a` and `n : ℕ`,
 `degree_zero_part.deg` picks the numerator `a`
 -/
-def degree_zero_part.num {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) : A :=
+def degree_zero_part.num {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) : A :=
 x.2.some_spec.some.1
 
-lemma degree_zero_part.num_mem {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) :
-  degree_zero_part.num f_deg x ∈ 𝒜 (m * degree_zero_part.deg f_deg x) :=
+lemma degree_zero_part.num_mem {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) :
+  degree_zero_part.num x ∈ 𝒜 (m * degree_zero_part.deg x) :=
 x.2.some_spec.some.2
 
-lemma degree_zero_part.eq {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x : A⁰_ f_deg) :
-  x.1 = mk (degree_zero_part.num f_deg x) ⟨f^(degree_zero_part.deg f_deg x), ⟨_, rfl⟩⟩ :=
+lemma degree_zero_part.eq {f : A} {m : ℕ} {f_deg : f ∈ 𝒜 m} (x : A⁰_ f_deg) :
+  x.1 = mk (degree_zero_part.num x) ⟨f^(degree_zero_part.deg x), ⟨_, rfl⟩⟩ :=
 x.2.some_spec.some_spec
 
 lemma degree_zero_part.mul_val {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) (x y : A⁰_ f_deg) :
   (x * y).1 = x.1 * y.1 := rfl
 
 end
 
+section clear_denominator
+
+-- this is a wrapper around `is_localization.exist_integer_multiples_of_finset`, the main purpose
+-- of this lemma is to make the degree of denominator explicit.
+private lemma clear_denominator {f : A} (s : finset (away f)) :
+  ∃ (n : ℕ), ∀ (x : away f), x ∈ s →
+    x * (mk (f^n) 1 : away f) ∈
+    (λ y, (mk y 1 : localization.away f)) '' set.univ :=
+begin
+  rcases is_localization.exist_integer_multiples_of_finset (submonoid.powers f) s with
+    ⟨⟨_, ⟨n, rfl⟩⟩, h⟩,
+  refine ⟨n, λ x hx, _⟩,
+  rcases h x hx with ⟨a, eq1⟩,
+  induction x using localization.induction_on with data,
+  rcases data with ⟨x, y⟩,
+  change mk a 1 = f^n • _ at eq1,
+  unfold has_scalar.smul localization.smul at eq1,
+  rw [localization.lift_on_mk, smul_eq_mul] at eq1,
+  rw [mk_mul, mul_one, mul_comm, ← eq1],
+  refine ⟨a, trivial, rfl⟩,
```

Comment:
```suggestion
  rw [mk_one_eq_algebra_map, set.image_univ, mul_comm, ← algebra.smul_def],
  exact h _ hx,
```

I'm not sure if this needs to be a separate lemma now that the proof is four lines long?